docs/notary.md

```diff
@@ -0,0 +1,535 @@
+chain environment. The on-chain P2P service is aimed to automate, accelerate and
+secure the process of signature collection. The service was initially designed as
+a solution for
+[multisignature transaction forming](https://github.com/neo-project/neo/issues/1573#issue-600384746)
+and described in the [proposal](https://github.com/neo-project/neo/issues/1573#issuecomment-704874472).
+
+The original problem definition:
+> Several parties want to sign one transaction, it can either be a set of signatures
+> for multisignature signer or multiple signers in one transaction. It's assumed
+> that all parties can generate the same transaction (with the same hash) without
+> any interaction, which is the case for oracle nodes or NeoFS inner ring nodes.
+> 
+> As some of the services using this mechanism can be quite sensitive to the
+> latency of their requests processing it should be possible to construct complete
+> transaction within the time frame between two consecutive blocks.
+
+
+## Components and functionality
+The service consists of a native contract and a node module. Native contract is
+mostly concerned with verification, fees and payment guarantees, while module is
+doing the actual work. It uses generic `Conflicts` and `NotValidBefore`
+transaction attributes for its purposes as well as an additional special one
+(`Notary assisted`).
+
+A new designated role is added, `P2PNotary`. It can have arbitrary number of
+keys associated with it.
+
+Using the service costs some GAS, so below we operate with `FEE` as a unit of cost
+for this service. `FEE` is set to be 0.1 GAS.
+
+### Transaction attributes
+
+#### Conflicts
+
+This attribute makes the chain only accept one transaction of the two conflicting
+and adds an ability to give a priority to any of the two if needed. This
+attribute was originally proposed in
+[neo-project/neo#1991](https://github.com/neo-project/neo/issues/1991).
+
+The attribute has Uint256 data inside of it containing the hash of conflicting
+transaction. It is allowed to have multiple attributes of this type.
+
+#### NotValidBefore
+
+This attribute makes transaction invalid before certain height. This attribute
+was originally proposed in
+[neo-project/neo#1992](https://github.com/neo-project/neo/issues/1992).
+
+The attribute has uint32 data inside which is the block height starting from
+which the transaction is considered to be valid. It can be seen as the opposite
+of `ValidUntilBlock`, using both allows to have a window of valid block numbers
+that this transaction could be accepted into. Transactions with this attribute
+are not accepted into mempool before specified block is persisted.
+
+It can be used to create some transactions in advance with a guarantee that they
+won't be accepted until specified block.
+
+#### NotaryAssisted
+
+This attribute contains one byte containing the number of transactions collected
+by the service. It could be 0 for fallback transaction or `NKeys` for normal
```

Comment:
shouldn't `NKeys` term be defined somewhere? or it is supposed that the reader has read the original issue and proposal?

Reply:
Fixed.